File: src/dsp/polyphase_resampler.cpp
#include "polyphase_resampler.h"

#include <algorithm>
#include <cmath>
#include <stdexcept>

#include "filter_design.h"

namespace jalert::dsp {

PolyphaseResampler::PolyphaseResampler(double input_rate_hz, double output_rate_hz,
                                       double cutoff_hz, int nphases)
    : nphases_(nphases) {
    if (input_rate_hz <= 0.0) throw std::invalid_argument("input_rate_hz must be positive");
    if (output_rate_hz <= 0.0) throw std::invalid_argument("output_rate_hz must be positive");

    // Cut-off must stay below 0.45×min(Nyquist) on both sides to avoid aliasing /
    // first-image leakage.
    const double fc = std::min(cutoff_hz, 0.45 * std::min(input_rate_hz, output_rate_hz));

    // Prototype span ~1.5 ms of input — gentle skirt — but capped so the
    // per-output convolution stays cheap at the MHz-class rates this plugin
    // runs (1.5 ms × 2 MHz would be 3000 taps/arm). With nphases arms the
    // capped prototype is still nphases × taps_per_arm taps, plenty for a clean
    // anti-alias well below the 250 kHz channel cutoff.
    taps_per_arm_ = std::max(8, static_cast<int>(std::round(1.5e-3 * input_rate_hz)));
    taps_per_arm_ = std::min(taps_per_arm_, 32);
    half_taps_ = taps_per_arm_ / 2;

    const int proto_len = nphases_ * taps_per_arm_;
    const int odd_len = (proto_len & 1) == 0 ? proto_len + 1 : proto_len;
    std::vector<float> proto_odd = lowpass_taps(odd_len, static_cast<float>(fc),
                                                static_cast<float>(nphases * input_rate_hz));
    std::vector<float> proto(proto_len);
    std::copy(proto_odd.begin(), proto_odd.begin() + proto_len, proto.begin());
    for (int i = 0; i < proto_len; ++i) proto[i] *= nphases_;

    arms_.resize(nphases_);
    for (int a = 0; a < nphases_; ++a) {
        std::vector<float> arm(taps_per_arm_);
        for (int j = 0; j < taps_per_arm_; ++j) arm[j] = proto[a + j * nphases_];
        std::reverse(arm.begin(), arm.end());
        arms_[a] = std::move(arm);
    }

    step_ = input_rate_hz / output_rate_hz;
    buf_len_ = taps_per_arm_ + 16;
    buf_re_.assign(buf_len_, 0.0f);
    buf_im_.assign(buf_len_, 0.0f);
    reset();
}

void PolyphaseResampler::reset() {
    std::fill(buf_re_.begin(), buf_re_.end(), 0.0f);
    std::fill(buf_im_.begin(), buf_im_.end(), 0.0f);
    buf_start_ = 0;
    buf_count_ = 0;
    in_index_ = 0;
    next_out_ = static_cast<double>(half_taps_);
}

void PolyphaseResampler::step(Complex32 z, std::vector<Complex32>& output) {
    // Push into the ring, dropping the oldest if it is full.
    if (buf_count_ == buf_len_) {
        buf_start_ = (buf_start_ + 1 == buf_len_) ? 0 : buf_start_ + 1;
        --buf_count_;
    }
    int w = buf_start_ + buf_count_;
    if (w >= buf_len_) w -= buf_len_;
    buf_re_[w] = z.re;
    buf_im_[w] = z.im;
    ++buf_count_;
    ++in_index_;

    const long long oldest = in_index_ - buf_count_;
    const long long newest = in_index_ - 1;

    while (true) {
        long long ip = static_cast<long long>(std::floor(next_out_));
        double frac = next_out_ - ip;
        int arm_idx = static_cast<int>(std::round(frac * nphases_));
        if (arm_idx >= nphases_) { arm_idx = 0; ++ip; }

        const long long lo = ip - half_taps_;
        const long long hi = lo + taps_per_arm_ - 1;
        if (hi > newest) break;
        if (lo < oldest) {
            // Window underflowed the buffer — only possible if the ring is too
            // small for the ratio. Skip rather than read stale data.
            next_out_ += step_;
            continue;
        }

        const std::vector<float>& arm = arms_[arm_idx];
        int bi = static_cast<int>(buf_start_ + (lo - oldest));
        if (bi >= buf_len_) bi -= buf_len_;
        float yr = 0.0f, yi = 0.0f;
        for (int j = 0; j < taps_per_arm_; ++j) {
            const float t = arm[j];
            yr += buf_re_[bi] * t;
            yi += buf_im_[bi] * t;
            if (++bi == buf_len_) bi = 0;
        }
        output.emplace_back(yr, yi);
        next_out_ += step_;
    }
}

} // namespace jalert::dsp

File: SDRSharp.JAlert/Dsp/PolyphaseResampler.cs
// Arbitrary-ratio polyphase resampler. Used to re-pin the input IQ stream to the
// J-ALERT working rate (4 sps × 256 ksym/s = 1.024 MHz).
// Ported from src/dsp/polyphase_resampler.{h,cpp}.

using System;
using System.Collections.Generic;

namespace SDRSharp.JAlert.Dsp
{
    internal sealed class PolyphaseResampler
    {
        private readonly int _nphases;
        private readonly int _tapsPerArm;
        private readonly int _halfTaps;
        private readonly float[][] _arms;
        private readonly double _step;
        private readonly float[] _bufRe;
        private readonly float[] _bufIm;
        private readonly int _bufLen;
        private int _bufStart;
        private int _bufCount;
        private long _inIndex;
        private double _nextOut;

        public PolyphaseResampler(double inputRateHz, double outputRateHz, double cutoffHz, int nphases = 32)
        {
            if (inputRateHz <= 0.0) throw new ArgumentException("inputRateHz must be positive");
            if (outputRateHz <= 0.0) throw new ArgumentException("outputRateHz must be positive");
            _nphases = nphases;

            // Cut-off must stay below 0.45×min(Nyquist) on both sides to avoid
            // aliasing / first-image leakage.
            double fc = Math.Min(cutoffHz, 0.45 * Math.Min(inputRateHz, outputRateHz));

            // Prototype span ~1.5 ms of input — gentle skirt — but capped so the
            // per-output convolution stays cheap at the MHz-class rates this plugin
            // runs.
            _tapsPerArm = Math.Max(8, (int)Math.Round(1.5e-3 * inputRateHz));
            _tapsPerArm = Math.Min(_tapsPerArm, 32);
            _halfTaps = _tapsPerArm / 2;

            int protoLen = _nphases * _tapsPerArm;
            int oddLen = (protoLen & 1) == 0 ? protoLen + 1 : protoLen;
            float[] protoOdd = FilterDesign.LowpassTaps(oddLen, (float)fc, (float)(nphases * inputRateHz));
            float[] proto = new float[protoLen];
            Array.Copy(protoOdd, proto, protoLen);
            for (int i = 0; i < protoLen; ++i) proto[i] *= _nphases;

            _arms = new float[_nphases][];
            for (int a = 0; a < _nphases; ++a)
            {
                float[] arm = new float[_tapsPerArm];
                for (int j = 0; j < _tapsPerArm; ++j) arm[j] = proto[a + j * _nphases];
                Array.Reverse(arm);
                _arms[a] = arm;
            }

            _step = inputRateHz / outputRateHz;
            _bufLen = _tapsPerArm + 16;
            _bufRe = new float[_bufLen];
            _bufIm = new float[_bufLen];
            Reset();
        }

        public void Reset()
        {
            Array.Clear(_bufRe, 0, _bufLen);
            Array.Clear(_bufIm, 0, _bufLen);
            _bufStart = 0;
            _bufCount = 0;
            _inIndex = 0;
            _nextOut = _halfTaps;
        }

        // Push one input sample; append zero or more output samples to output.
        public void Step(Complex32 z, List<Complex32> output)
        {
            // Push into the ring, dropping the oldest if it is full.
            if (_bufCount == _bufLen)
            {
                _bufStart = (_bufStart + 1 == _bufLen) ? 0 : _bufStart + 1;
                --_bufCount;
            }
            int w = _bufStart + _bufCount;
            if (w >= _bufLen) w -= _bufLen;
            _bufRe[w] = z.Re;
            _bufIm[w] = z.Im;
            ++_bufCount;
            ++_inIndex;

            long oldest = _inIndex - _bufCount;
            long newest = _inIndex - 1;

            while (true)
            {
                long ip = (long)Math.Floor(_nextOut);
                double frac = _nextOut - ip;
                int armIdx = (int)Math.Round(frac * _nphases);
                if (armIdx >= _nphases) { armIdx = 0; ++ip; }

                long lo = ip - _halfTaps;
                long hi = lo + _tapsPerArm - 1;
                if (hi > newest) break;
                if (lo < oldest)
                {
                    // Window underflowed the buffer — only possible if the ring is
                    // too small for the ratio. Skip rather than read stale data.
                    _nextOut += _step;
                    continue;
                }

                float[] arm = _arms[armIdx];
                int bi = (int)(_bufStart + (lo - oldest));
                if (bi >= _bufLen) bi -= _bufLen;
                float yr = 0.0f, yi = 0.0f;
                for (int j = 0; j < _tapsPerArm; ++j)
                {
                    float t = arm[j];
                    yr += _bufRe[bi] * t;
                    yi += _bufIm[bi] * t;
                    if (++bi == _bufLen) bi = 0;
                }
                output.Add(new Complex32(yr, yi));
                _nextOut += _step;
            }
        }
    }
}
